## Final Project Submission

Please fill out:
* Student name: Richard Gachiri Muriithi
* Student pace: part time 
* Scheduled project review date/time: 
* Instructor name: 
* Blog post URL:


**Loading data**

We load data into a data structure called a **dataframe**. A dataframe contains rows and columns; it can be easily manipulated hence appropriate for data analysis.

In [44]:
# Your code here - remember to use markdown cells for comments as well!
import numpy as np
import pandas as pd
import seaborn as sns

In [45]:
df = pd.read_csv('imdb.title.ratings.csv.gz')
df.head() # checks the first 5 rows

tconst  averagerating  numvotes
0  tt10356526            8.3        31
1  tt10384606            8.9       559
2   tt1042974            6.4        20
3   tt1043726            4.2     50352
4   tt1060240            6.5        21

In [46]:
df1 = pd.read_csv('imdb.title.basics.csv.gz')
df1.head() # checks the first 5 rows

tconst                    primary_title              original_title  \
0  tt0063540                        Sunghursh                   Sunghursh   
1  tt0066787  One Day Before the Rainy Season             Ashad Ka Ek Din   
2  tt0069049       The Other Side of the Wind  The Other Side of the Wind   
3  tt0069204                  Sabse Bada Sukh             Sabse Bada Sukh   
4  tt0100275         The Wandering Soap Opera       La Telenovela Errante   

   start_year  runtime_minutes                genres  
0        2013            175.0    Action,Crime,Drama  
1        2019            114.0       Biography,Drama  
2        2018            122.0                 Drama  
3        2018              NaN          Comedy,Drama  
4        2017             80.0  Comedy,Drama,Fantasy

In [47]:
df2 = pd.read_csv('bom.movie_gross.csv.gz')
df2.head() # checks the first 5 rows

title studio  domestic_gross  \
0                                  Toy Story 3     BV     415000000.0   
1                   Alice in Wonderland (2010)     BV     334200000.0   
2  Harry Potter and the Deathly Hallows Part 1     WB     296000000.0   
3                                    Inception     WB     292600000.0   
4                          Shrek Forever After   P/DW     238700000.0   

  foreign_gross  year  
0     652000000  2010  
1     691300000  2010  
2     664300000  2010  
3     535700000  2010  
4     513900000  2010

In [48]:
df.info() # checks for the overview of the data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 73856 entries, 0 to 73855
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   tconst         73856 non-null  object 
 1   averagerating  73856 non-null  float64
 2   numvotes       73856 non-null  int64  
dtypes: float64(1), int64(1), object(1)
memory usage: 1.7+ MB


In [5]:
df1.info() # checks for the overview of the data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 146144 entries, 0 to 146143
Data columns (total 6 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   tconst           146144 non-null  object 
 1   primary_title    146144 non-null  object 
 2   original_title   146123 non-null  object 
 3   start_year       146144 non-null  int64  
 4   runtime_minutes  114405 non-null  float64
 5   genres           140736 non-null  object 
dtypes: float64(1), int64(1), object(4)
memory usage: 6.7+ MB


In [49]:
df2.info() # checks for the overview of the data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3387 entries, 0 to 3386
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   title           3387 non-null   object 
 1   studio          3382 non-null   object 
 2   domestic_gross  3359 non-null   float64
 3   foreign_gross   2037 non-null   object 
 4   year            3387 non-null   int64  
dtypes: float64(1), int64(1), object(3)
memory usage: 132.4+ KB


In [5]:
df.shape # checks for the number of rows and columns

(73856, 3)

In [7]:
df1.shape # checks for the number of rows and columns

(146144, 6)

In [8]:
df2.shape # checks for the number of rows and columns

(3387, 5)

In [32]:
df.dtypes # dtype attribute checks the typer of data 

tconst            object
averagerating    float64
numvotes           int64
dtype: object

In [33]:
df1.dtypes # dtype attribute checks the typer of data 

tconst              object
primary_title       object
original_title      object
start_year           int64
runtime_minutes    float64
genres              object
dtype: object

In [34]:
df2.dtypes # dtype attribute checks the typer of data 

title              object
studio             object
domestic_gross    float64
foreign_gross      object
year                int64
dtype: object

In [35]:
df.describe() # check for a statistical summary of the data

averagerating      numvotes
count   73856.000000  7.385600e+04
mean        6.332729  3.523662e+03
std         1.474978  3.029402e+04
min         1.000000  5.000000e+00
25%         5.500000  1.400000e+01
50%         6.500000  4.900000e+01
75%         7.400000  2.820000e+02
max        10.000000  1.841066e+06

In [36]:
df1.describe() # check for a statistical summary of the data

start_year  runtime_minutes
count  146144.000000    114405.000000
mean     2014.621798        86.187247
std         2.733583       166.360590
min      2010.000000         1.000000
25%      2012.000000        70.000000
50%      2015.000000        87.000000
75%      2017.000000        99.000000
max      2115.000000     51420.000000

In [37]:
df2.describe() # check for a statistical summary of the data

domestic_gross         year
count    3.359000e+03  3387.000000
mean     2.874585e+07  2013.958075
std      6.698250e+07     2.478141
min      1.000000e+02  2010.000000
25%      1.200000e+05  2012.000000
50%      1.400000e+06  2014.000000
75%      2.790000e+07  2016.000000
max      9.367000e+08  2018.000000

### Dealing with missing values

We will check first for missing data values  in all the datasets.
There are various methods of dealing with missing values:

1. Drop missing values
2. Fill/replace with mean/median/mode/, backward fill/forward fill, certain value
3. Interpolate

In [50]:
df.isnull().sum() # Gets count of missing values, imdb.title.ratings.csv

tconst           0
averagerating    0
numvotes         0
dtype: int64

We can see that from the dataset, imdb.title.ratings they is no missing values.

In [51]:
df1.isnull().sum() # Gets count of missing values, imdb.title.basics.csv

tconst                 0
primary_title          0
original_title        21
start_year             0
runtime_minutes    31739
genres              5408
dtype: int64

We can see that from the dataset, imdb.title.basics they is  missing values in:
1. original_title, 21 missing values
2. runtime_minutes, 31739 missing values
3. genres, 5408 missing values. 


In [52]:
df2.isnull().sum() # Gets count of missing values, born.movie_gross.cvs

title                0
studio               5
domestic_gross      28
foreign_gross     1350
year                 0
dtype: int64

We can see that from the dataset, born.movie_gross.cvs they is  missing values in:
1. studio, 5 missing values
2. domestic_gross, 28 missing values
3. foreign_gross, 1350 missing values. 

In [53]:
df.isna().sum() # Gets count of missing values

tconst           0
averagerating    0
numvotes         0
dtype: int64

In [54]:
df1.isna().sum()# Gets count of missing values

tconst                 0
primary_title          0
original_title        21
start_year             0
runtime_minutes    31739
genres              5408
dtype: int64

In [62]:
df2.isna().sum() # Gets count of missing values

title                0
studio               5
domestic_gross      28
foreign_gross     1350
year                 0
dtype: int64

In [55]:
missing_values = df.isna() # Checks missing values
missing_values.head(50)

tconst  averagerating  numvotes
0    False          False     False
1    False          False     False
2    False          False     False
3    False          False     False
4    False          False     False
5    False          False     False
6    False          False     False
7    False          False     False
8    False          False     False
9    False          False     False
10   False          False     False
11   False          False     False
12   False          False     False
13   False          False     False
14   False          False     False
15   False          False     False
16   False          False     False
17   False          False     False
18   False          False     False
19   False          False     False
20   False          False     False
21   False          False     False
22   False          False     False
23   False          False     False
24   False          False     False
25   False          False     False
26   False          False     False
27   False          False     False
28   False          False     False
29   False          False     False
30   False          False     False
31   False          False     False
32   False          False     False
33   False          False     False
34   False          False     False
35   False          False     False
36   False          False     False
37   False          False     False
38   False          False     False
39   False          False     False
40   False          False     False
41   False          False     False
42   False          False     False
43   False          False     False
44   False          False     False
45   False          False     False
46   False          False     False
47   False          False     False
48   False          False     False
49   False          False     False

In [26]:
missing_values = df1.isna() # Checks missing values
missing_values.head(50)

tconst  primary_title  original_title  start_year  runtime_minutes  genres
0    False          False           False       False            False   False
1    False          False           False       False            False   False
2    False          False           False       False            False   False
3    False          False           False       False             True   False
4    False          False           False       False            False   False
5    False          False           False       False            False   False
6    False          False           False       False             True   False
7    False          False           False       False            False   False
8    False          False           False       False             True   False
9    False          False           False       False            False   False
10   False          False           False       False            False   False
11   False          False           False       False            False   False
12   False          False           False       False            False   False
13   False          False           False       False            False   False
14   False          False           False       False            False   False
15   False          False           False       False            False   False
16   False          False           False       False             True    True
17   False          False           False       False            False   False
18   False          False           False       False            False   False
19   False          False           False       False            False   False
20   False          False           False       False            False   False
21   False          False           False       False             True   False
22   False          False           False       False            False    True
23   False          False           False       False            False   False
24   False          False           False       False            False   False
25   False          False           False       False            False   False
26   False          False           False       False             True   False
27   False          False           False       False            False   False
28   False          False           False       False            False   False
29   False          False           False       False            False   False
30   False          False           False       False            False   False
31   False          False           False       False             True   False
32   False          False           False       False            False   False
33   False          False           False       False             True   False
34   False          False           False       False             True   False
35   False          False           False       False            False    True
36   False          False           False       False            False   False
37   False          False           False       False            False   False
38   False          False           False       False            False   False
39   False          False           False       False            False   False
40   False          False           False       False            False    True
41   False          False           False       False            False   False
42   False          False           False       False            False   False
43   False          False           False       False            False   False
44   False          False           False       False             True    True
45   False          False           False       False            False    True
46   False          False           False       False            False   False
47   False          False           False       False            False   False
48   False          False           False       False            False   False
49   False          False           False       False 

In [12]:
df2.isna().sum()

title                0
studio               5
domestic_gross      28
foreign_gross     1350
year                 0
dtype: int64

In [25]:
missing_values = df2.isna() # Checks missing values
missing_values.tail(50)

title  studio  domestic_gross  foreign_gross   year
3337  False   False           False           True  False
3338  False   False           False           True  False
3339  False   False           False           True  False
3340  False   False           False           True  False
3341  False   False           False           True  False
3342  False   False           False          False  False
3343  False   False           False           True  False
3344  False   False           False           True  False
3345  False   False           False           True  False
3346  False   False           False           True  False
3347  False   False           False           True  False
3348  False   False           False           True  False
3349  False   False           False           True  False
3350  False   False           False           True  False
3351  False   False           False           True  False
3352  False   False           False           True  False
3353  False   False           False          False  False
3354  False   False           False           True  False
3355  False   False           False           True  False
3356  False   False           False           True  False
3357  False   False           False           True  False
3358  False   False           False           True  False
3359  False   False           False           True  False
3360  False   False           False           True  False
3361  False   False           False           True  False
3362  False   False           False           True  False
3363  False   False           False           True  False
3364  False   False           False           True  False
3365  False   False           False           True  False
3366  False   False           False           True  False
3367  False   False           False           True  False
3368  False   False           False           True  False
3369  False   False           False           True  False
3370  False   False           False           True  False
3371  False   False           False           True  False
3372  False   False           False           True  False
3373  False   False           False           True  False
3374  False   False           False           True  False
3375  False   False           False           True  False
3376  False   False           False           True  False
3377  False   False           False           True  False
3378  False   False           False           True  False
3379  False   False           False           True  False
3380  False   False           False           True  False
3381  False   False           False           True  False
3382  False   False           False           True  False
3383  False   False           False           True  False
3384  False   False           False           True  False
3385  False   False           False           True  False
3386  False   False           False           True  False

Below we will check the percentage of missing value of the datasets **imdb.title.ratings, imdb.title.basic and born.movie_gross**. 

This will help use in knowing how to deal with the missing values in the datasets provided. 

In [58]:
def missing_values(data):
    # identify the total missing values per column
    # sort in order
    miss = data.isnull().sum().sort_values(ascending = False)

    # calculate percentage of the missing values
    percentage_miss = (data.isnull().sum() / len(data)).sort_values(ascending = False)

    # store in a dataframe
    missing = pd.DataFrame({"Missing Values": miss, "Percentage": percentage_miss}).reset_index()

    # remove values that are missing
    missing.drop(missing[missing["Percentage"] == 0].index, inplace = True)

    return missing

In [59]:
missing_data = missing_values(df)
missing_data

Empty DataFrame
Columns: [index, Missing Values, Percentage]
Index: []

In [60]:
missing_data = missing_values(df1)
missing_data

index  Missing Values  Percentage
0  runtime_minutes           31739    0.217176
1           genres            5408    0.037005
2   original_title              21    0.000144

In [61]:
missing_data = missing_values(df2)
missing_data

index  Missing Values  Percentage
0   foreign_gross            1350    0.398583
1  domestic_gross              28    0.008267
2          studio               5    0.001476